Content.Shared/Doors/Systems/TurnstileSystem.cs

```diff
@@ -0,0 +1,109 @@
+using Content.Shared.Access.Systems;
+using Content.Shared.Doors.Components;
+using Content.Shared.Movement.Pulling.Systems;
+using Content.Shared.Popups;
+using Content.Shared.Whitelist;
+using Robust.Shared.Animations;
+using Robust.Shared.Audio.Systems;
+using Robust.Shared.Network;
+using Robust.Shared.Physics.Events;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Doors.Systems;
+
+/// <summary>
+/// This handles logic and interactions related to <see cref="TurnstileComponent"/>
+/// </summary>
+public sealed class TurnstileSystem : EntitySystem
+{
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly INetManager _net = default!;
+    [Dependency] private readonly AccessReaderSystem _accessReader = default!;
+    [Dependency] private readonly SharedAnimationPlayerSystem _animationPlayer = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly EntityWhitelistSystem _entityWhitelist = default!;
+    [Dependency] private readonly PullingSystem _pulling = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<TurnstileComponent, PreventCollideEvent>(OnPreventCollide);
+        SubscribeLocalEvent<TurnstileComponent, EndCollideEvent>(OnEndCollide);
+    }
+
+    private void OnPreventCollide(Entity<TurnstileComponent> ent, ref PreventCollideEvent args)
+    {
+        if (args.Cancelled || !args.OurFixture.Hard)
+            return;
+
+        if (ent.Comp.CollideExceptions.Contains(args.OtherEntity))
+        {
+            // We need to add this in here too for chain pulls
+            if (_pulling.GetPulling(args.OtherEntity) is { } uid &&
+                ent.Comp.CollideExceptions.Add(uid))
+            {
+                Dirty(ent);
+            }
+
+            args.Cancelled = true;
+            return;
+        }
+
+        // unblockables go through for free.
+        if (_entityWhitelist.IsWhitelistFail(ent.Comp.ProcessWhitelist, args.OtherEntity))
+        {
+            args.Cancelled = true;
+            return;
+        }
+
+        if (_timing.CurTime < ent.Comp.NextPassTime)
+            return;
+
```

Comment:
Isn't this going to mispredict if more than 1 ent even touches it. That seems pretty bad? Movement stuff should always be fully predicted. Worst case the client just tells the server if it got rejected or not just so it doesn't feel bad ingame.